packages/platforms/celestica/x86-64/x86-64-cel-redstone-xp/onlp/builds/src/module/src/i2c_dev.c

```diff
@@ -0,0 +1,483 @@
+	close(fd);
+	return ret;
+}
+//obsoleted
+int i2c_smbus_write_16reg(int busNum, short addr, short reg,
+									int size, const __u8 *buf)
+{
+	int i;
+	int fd;
+	int ret = -1;
+	char wBuf[3];
+	char fname[20];
+	struct i2c_rdwr_ioctl_data ioctl_data;
+	struct i2c_msg msg;
+
+	sprintf(fname, "/dev/i2c-%d", busNum);
+	fd = open(fname, O_RDWR);
+
+    if(fd < 0)
+		return ret;
+	ret = ioctl(fd, I2C_SLAVE, addr);
+
+    for(i=0; i<size; i++) {
+		wBuf[1] = (reg + i) & 0x0FF;
+		wBuf[0] = ((reg + i) >> 8) & 0x0FF;
+		wBuf[2] = buf[i];
+		msg.addr= addr;
+		msg.flags = 0;
+		msg.len= 3;
+		msg.buf= wBuf;
+		ioctl_data.nmsgs= 1;
+		ioctl_data.msgs= &msg;
+		ret = ioctl(fd, I2C_RDWR, &ioctl_data);
+		if(ret)
+			break;
+	}
+	close(fd);
+	return ret;
+}
+
+int i2c_smbus_write_16reg_8byte(int busNum, short addr, short reg,
+									int size, const __u8 *buf)
+{
+	int i;
+	int fd;
+	int ret = -1;
+	char wBuf[10];
+	char fname[20];
+	struct i2c_rdwr_ioctl_data ioctl_data;
+	struct i2c_msg msg;
+
+    sprintf(fname, "/dev/i2c-%d", busNum);
+	fd = open(fname, O_RDWR);
+
+    if(fd < 0)
+		return ret;
+
+    for(i=0; i<size; i+=8) {
+		wBuf[1] = (reg + i) & 0x0FF;
+		wBuf[0] = ((reg + i) >> 8) & 0x0FF;
+		memcpy(&wBuf[2],&buf[i], 8);
+		msg.addr= addr;
+		msg.flags = 0;
+		msg.len= 10+1;
+		msg.buf= wBuf;
+		ioctl_data.nmsgs= 1;
+		ioctl_data.msgs= &msg;
+		ret = ioctl(fd, I2C_RDWR, &ioctl_data);
+		if(ret)
+			break;
+	}
+	close(fd);
+	return ret;
+}
+
+int chips_read_byte(char addr, __u8 reg, __u8 *data)
+{
+	char busNum = 1;
+	dataType type = LEN_BYTE;
+	int ret = -1;
+	ret = i2_smbus_read(busNum, addr, reg, type, 1, data);
+	return ret;
+}
+
+int chips_read_word(char addr, __u8 reg, unsigned short *data)
+{
+	int busNum = 1;
+	unsigned char buf[2];
+	dataType type = LEN_WORD;
+	int ret = -1;
+	ret = i2_smbus_read(busNum, addr, reg, type, 1, buf);
+	if(ret >= 0)
+	{
+		*data = ((unsigned short)buf[1] << 8) | buf[0];
+	}
+	return ret;
+}
+
+int
+chips_write_word(char addr, __u8 reg, unsigned short data)
+{
+	int busNum = 1;
+	dataType type = LEN_WORD;
+	int ret = -1;
+	unsigned char buf[2] =
+		{ (data >> 8) & 0x00ff, data & 0x00ff};
+	ret = i2c_smbus_write(busNum, addr, reg, type, 1, (unsigned char *)&buf);
+	//ret = i2c_smbus_write(busNum, addr, reg, type, 1, (char *)&data);
+	return ret;
+}
+
+int
+chips_write_word_pec(char addr, __u8 reg, unsigned short data, int pec_flag)
+{
+	int busNum = 1;
+	dataType type = LEN_WORD;
+	int ret = -1;
+	ret = i2c_smbus_write_pec(busNum, addr, reg, type, pec_flag, (unsigned char *)&data);
+	return ret;
+}
+
+int
+chips_write_block(char addr, __u8 reg, int byteCnt, unsigned char* wrData)
+{
+	int busNum = 1;
+	int ret = -1;
+
+	ret = i2c_smbus_write_block_data(busNum, addr, reg, byteCnt, wrData);
+	return ret;
+}
+
+int
+chips_write_byte(char addr, __u8 reg, const __u8 data)
+{
+	int busNum = 1;
+	dataType type = LEN_BYTE;
+	int ret = -1;
+	ret = i2c_smbus_write(busNum, addr, reg, type, 1, (unsigned char *)&data);
```

Comment:
While you are free to reimplement all of the userspace I2C access please take a look at the available userspace I2C device management APIs already available in the onlplib module. 

Declaring and accessing i2c device trees (mux trees, device addresses, etc) are all defined with available infrastructure here. You don't need to rewrite all of the hierarchies. You might consider using it for future implementations, it really does simplify your userspace implementation:
https://github.com/opencomputeproject/OpenNetworkLinux/blob/master/packages/base/any/onlp/src/onlplib/module/inc/onlplib/i2c.h